### PR TITLE
Add new resource alicloud_gpdb_elastic_instance and no longer supports to create new alicloud_gpdb_instance resource

### PR DESCRIPTION
**Pull request by jokerhenson**

### CODE REVIEW

File: alicloud/resource_alicloud_gpdb_elastic_instance.go [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice([]string{"VPC"}, false),
+			},
+			"payment_type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice([]string{"Postpaid", "Prepaid"}, false),
```
Comment (liuri1205) [inline alicloud/resource_alicloud_gpdb_elastic_instance.go:77]: `payment_type` 属性参考 `alicloud_tsdb_instance` 资源的方式处理下

File: alicloud/resource_alicloud_gpdb_elastic_instance.go [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+		}
+		return WrapError(err)
+	}
+	d.Set("engine", instance["Engine"])
+	d.Set("engine_version", instance["EngineVersion"])
+	d.Set("storage_type", instance["StorageType"])
+	d.Set("seg_node_num", instance["SegNodeNum"])
+	d.Set("storage_size", instance["StorageSize"])
+	d.Set("payment_type", instance["PayType"])
+	d.Set("instance_spec", convertDBInstanceClassToInstanceSpec(instance["DBInstanceClass"].(string)))
```
Comment (liuri1205) [inline alicloud/resource_alicloud_gpdb_elastic_instance.go:199]: `instance_spec` 属性查询接口要支持返回, 不能通过 `DBInstanceClass` 去特殊处理获取

File: website/docs/r/gpdb_instance.html.markdown [hunk 3]
```diff
@@ -19,6 +19,9 @@ You can see detail product introduction [here](https://www.alibabacloud.com/help
 
 -> **NOTE:**  Create instance or change instance would cost 10~15 minutes. Please make full preparation.
 
+-> **NOTE:**  This resource is used to manage a Reserved Storage Mode instance, and creating new reserved storage mode instance is no longer supported.
```
Comment (liuri1205) [inline website/docs/r/gpdb_instance.html.markdown:22]: 标明从哪个版本开始不再支持

File: website/docs/r/gpdb_elastic_instance.html.markdown [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  vswitch_id           = alicloud_vswitch.default.id
+}
+
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `engine` (Required) Database engine: `gpdb`.
```
Comment (liuri1205) [inline website/docs/r/gpdb_elastic_instance.html.markdown:59]: ForceNew 的属性也需要标识

File: alicloud/resource_alicloud_gpdb_elastic_instance.go [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+	}
+	if d.HasChange("description") {
+		action := "ModifyDBInstanceDescription"
+		request := map[string]interface{}{
+			"RegionId":              client.RegionId,
+			"DBInstanceId":          d.Id(),
+			"DBInstanceDescription": d.Get("description"),
+			"SourceIp":              client.SourceIp,
+		}
+		response, err := conn.DoRequest(StringPointer(action), nil, StringPointer("POST"), StringPointer("2016-05-03"), StringPointer("AK"), nil, request, &util.RuntimeOptions{})
```
Comment (liuri1205) [inline alicloud/resource_alicloud_gpdb_elastic_instance.go:230]: 使用 common 调用的接口, 对 err 都需要用 NeedRetry 函数判断进行重试

**Review by liuri1205 — CHANGES_REQUESTED** [review]

File: alicloud/resource_alicloud_gpdb_elastic_instance.go (hunk 2 above)
Comment (jokerhenson) [reply alicloud/resource_alicloud_gpdb_elastic_instance.go:199]: 先这么搞，下周api修改发布之后再把特殊处理去掉。

Comment (jokerhenson) [reply alicloud/resource_alicloud_gpdb_elastic_instance.go:199]: > `instance_spec` 属性查询接口要支持返回, 不能通过 `DBInstanceClass` 去特殊处理获取

先这么搞，下周api修改发布之后再把特殊处理去掉。

File: alicloud/resource_alicloud_gpdb_elastic_instance.go [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+		request["VSwitchId"] = vswitchId
+		if v, ok := request["ZoneId"].(string); !ok || v == "" {
+			request["ZoneId"] = vsw["ZoneId"]
+		}
+	}
+	runtime := util.RuntimeOptions{}
+	runtime.SetAutoretry(true)
+	request["ClientToken"] = buildClientToken("CreateECSDBInstance")
+	response, err = conn.DoRequest(StringPointer(action), nil, StringPointer("POST"), StringPointer("2016-05-03"), StringPointer("AK"), nil, request, &runtime)
+	if err != nil {
```
Comment (liuri1205) [inline alicloud/resource_alicloud_gpdb_elastic_instance.go:168]: NeedRetry 重试

File: alicloud/service_alicloud_gpdb.go [hunk 7]
```diff
@@ -45,6 +47,38 @@ func (s *GpdbService) DescribeGpdbInstance(id string) (instanceAttribute gpdb.DB
...
+	action := "DescribeDBInstanceOnECSAttribute"
+	request := map[string]interface{}{
+		"RegionId":     s.client.RegionId,
+		"DBInstanceId": id,
+		"SourceIp":     s.client.SourceIp,
+	}
+	var response map[string]interface{}
+	runtime := util.RuntimeOptions{}
+	runtime.SetAutoretry(true)
+	response, err = conn.DoRequest(StringPointer(action), nil, StringPointer("POST"), StringPointer("2016-05-03"), StringPointer("AK"), nil, request, &runtime)
```
Comment (liuri1205) [inline alicloud/service_alicloud_gpdb.go:64]: NeedRetry 重试

File: alicloud/resource_alicloud_gpdb_elastic_instance.go [hunk 8]
```diff
@@ -0,0 +1,323 @@
+package alicloud
+
+import (
+	"fmt"
+	util "github.com/alibabacloud-go/tea-utils/service"
```
Comment (liuri1205) [inline alicloud/resource_alicloud_gpdb_elastic_instance.go:5]: 运行 make fmt 格式化下代码

**Review by liuri1205 — CHANGES_REQUESTED** [review]

File: alicloud/resource_alicloud_gpdb_elastic_instance_test.go [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+					"storage_size":          "50",
+					"instance_spec":         "2C16G",
+					"description":           "tf-testAccGpdbInstance_6.0",
+					"instance_network_type": "VPC",
+					"payment_type":          "PayAsYouGo",
+					"vswitch_id":            "${local.vswitch_id}",
+				}),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheck(map[string]string{
+						"description": fmt.Sprintf("tf-testAccGpdbInstance_6.0"),
```
Comment (liuri1205) [inline alicloud/resource_alicloud_gpdb_elastic_instance_test.go:144]: 查询接口返回的属性都需要 Check 一下

**Review by liuri1205 — CHANGES_REQUESTED** [review]

File: alicloud/resource_alicloud_gpdb_elastic_instance.go [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+			"storage_type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice([]string{"cloud_essd", "cloud_efficiency"}, false),
+			},
+			"seg_node_num": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_gpdb_elastic_instance.go:51]: Does the seg_node_num has validate values or some limitations?

File: alicloud/resource_alicloud_gpdb_elastic_instance.go [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice([]string{"VPC"}, false),
+			},
+			"payment_type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice([]string{"PayAsYouGo", "Subscription"}, false),
+			},
+			"period": {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_gpdb_elastic_instance.go:80]: Please use the specific resource attribute `payment_duration_unit` instead `period`

File: alicloud/resource_alicloud_gpdb_elastic_instance.go [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice([]string{"Month", "Year"}, false),
+			},
+			"used_time": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice([]string{"1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12"}, false),
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_gpdb_elastic_instance.go:90]: Please use the specific resource attribute `payment_duration` instead `used_time`

File: alicloud/resource_alicloud_gpdb_elastic_instance.go [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+			},
+			"seg_node_num": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
+			},
+			"storage_size": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_gpdb_elastic_instance.go:56]: Does the `storage_size` has validate values or some limitations?

File: alicloud/resource_alicloud_gpdb_elastic_instance.go [hunk 14]
```diff
@@ -0,0 +1,336 @@
...
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"connection_string": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"security_ip_list": {
+				Type:     schema.TypeSet,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_gpdb_elastic_instance.go:112]: I think the `security_ip_list` type should be `schema.TypeList`.

File: alicloud/resource_alicloud_gpdb_elastic_instance.go [hunk 15]
```diff
@@ -0,0 +1,336 @@
...
+
+func resourceAlicloudGpdbElasticInstanceUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	gpdbService := GpdbService{client}
+	d.Partial(true)
+	conn, err := client.NewGpdbClient()
+	if err != nil {
+		return WrapError(err)
+	}
+	if d.HasChange("description") {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_gpdb_elastic_instance.go:235]: The description has been set in the Create API and there should check the ` d.IsNewResource() `

File: alicloud/resource_alicloud_gpdb_elastic_instance.go [hunk 16]
```diff
@@ -0,0 +1,336 @@
...
+		}
+		if err := gpdbService.ModifyGpdbSecurityIps(d.Id(), ipStr); err != nil {
+			return WrapError(err)
+		}
+		d.SetPartial("security_ip_list")
+	}
+	if d.IsNewResource() {
+		d.Partial(false)
+		return resourceAlicloudGpdbElasticInstanceRead(d, meta)
+	}
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_gpdb_elastic_instance.go:275]: The block 272-275 is needless and it should be removed.

File: alicloud/resource_alicloud_gpdb_elastic_instance_test.go [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+			{
+				Config: testAccConfig(map[string]interface{}{
+					"security_ip_list": []string{"10.168.1.12"},
+				}),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheck(map[string]string{
+						"security_ip_list.#":          "1",
+						"security_ip_list.4095458986": "10.168.1.12",
+					}),
+				),
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_gpdb_elastic_instance_test.go:183]: There missing the updating all attributes testcase.

File: alicloud/resource_alicloud_gpdb_instance.go [hunk 18]
```diff
@@ -120,39 +120,7 @@ func resourceAlicloudGpdbInstanceRead(d *schema.ResourceData, meta interface{})
...
-	}
-	d.SetId(response.DBInstanceId)
-
-	stateConf := BuildStateConf([]string{"Creating"}, []string{"Running"}, d.Timeout(schema.TimeoutCreate), 10*time.Minute, gpdbService.GpdbInstanceStateRefreshFunc(d.Id(), []string{"Deleting"}))
-
-	if _, err := stateConf.WaitForState(); err != nil {
-		return WrapErrorf(err, IdMsg, d.Id())
-	}
-	return resourceAlicloudGpdbInstanceUpdate(d, meta)
+	return Error("Sorry, creating new reserved storage mode instance is no longer supported since v1.127.0, you can use alicloud_gpdb_elastic_instance to create a flexible storage mode instance instead.")
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_gpdb_instance.go:123]: If the OpenAPI does not offline, there can not offline this resource. You can add the statement in its docs by a NOTE

File: alicloud/resource_alicloud_gpdb_instance_test.go [hunk 19]
```diff
@@ -108,7 +108,7 @@ func testSweepGpdbInstances(region string) error {
 	return nil
 }
 
-func TestAccAlicloudGpdbInstance_classic(t *testing.T) {
+func _TestAccAlicloudGpdbInstance_classic(t *testing.T) {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_gpdb_instance_test.go:111]: If you want to skip the testcase, you can add a prefix Skip in the testcase name.

**Review by xiaozhu36 — CHANGES_REQUESTED** [review]

**Pull request merged**